server/src/internalClusterTest/java/org/elasticsearch/action/support/master/TransportMasterNodeActionIT.java

```diff
@@ -78,22 +91,55 @@ public void testRoutingLoopProtection() {
                         connection.sendRequest(requestId, action, request, options);
                     }
                 });
-                mockTransportService.addRequestHandlingBehavior(ClusterStateAction.NAME, (handler, request, channel, task) -> {
-                    // assertThat(asInstanceOf(MasterNodeRequest.class, request).masterTerm(), equalTo(originalTerm)) TODO
-                    assertTrue("rerouted message received exactly once", reroutedMessageReceived.compareAndSet(false, true));
-                    handler.messageReceived(request, channel, task);
-                });
+
+                /*
+                 * Assert that no other node receives the re-routed message more than once, and only
+                 * from a node in the original term
+                 */
+                final var reroutedMessageReceived = new AtomicBoolean(false);
+                mockTransportService.addRequestHandlingBehavior(
+                    TransportClusterHealthAction.TYPE.name(),
+                    (handler, request, channel, task) -> {
+                        assertThat(asInstanceOf(MasterNodeRequest.class, request).masterTerm(), equalTo(originalTerm));
+                        assertTrue("rerouted message received exactly once", reroutedMessageReceived.compareAndSet(false, true));
+                        handler.messageReceived(request, channel, task);
+                    }
+                );
```

Comment:
I changed the test to use `TransportClusterHealthAction` as the redirected request instead of `ClusterStateAction` because `getMasterName` (used by `masterNodeClient` etc.) triggers a `ClusterStateAction`.